release_tools/notes.py

```diff
@@ -0,0 +1,261 @@
+#
+# Authors:
+#     Santiago Dueñas <sduenas@bitergia.com>
+#
+
+"""
+Script to generate the release notes of a package.
+
+It will read the changelog entries stored under 'releases/unreleased'
+directory and will generate a file with the list of changes.
+The script needs the name of the package and the version to release.
+"""
+
+import datetime
+import os
+import sys
+import textwrap
+
+import click
+
+from release_tools.entry import (CategoryChange,
+                                 determine_changelog_entries_dirpath,
+                                 read_changelog_entries)
+from release_tools.repo import GitHandler
+
+
+def validate_argument(ctx, param, value):
+    """Check argument valid values."""
+
+    value = value.strip("\n\r ")
+
+    if not value:
+        raise click.BadParameter("cannot be empty")
+    return value
+
+
+@click.command()
+@click.option('--dry-run', is_flag=True,
+              help="Do not write release notes file. Print to the standard output instead.")
+@click.option('--overwrite', is_flag=True,
+              help="Force to replace an existing release notes entry.")
+@click.argument('name', callback=validate_argument)
+@click.argument('version', callback=validate_argument)
+def notes(name, version, dry_run, overwrite):
+    """Generate release notes.
+
+    When you run this script, it will generate the release notes of the
+    package tracked by the current Git repository.
+
+    You will need to provide the 'NAME' of the package and the 'VERSION'
+    of the new release. The script will generate a Markdown document
+    under the 'releases' directory. Take into account the argument `NAME`
+    is only used as the title of the document.
+
+    In the case a release notes file for the same version already exists,
+    an error will be raised. Use '--overwrite' to force to replace the
+    existing file. If you just want to see the final result of the notes
+    but not generate a new file, please activate '--dry-run' flag.
+
+    NAME: title of the package for the release notes.
+
+    VERSION: version of the new release.
+    """
+    entry_list = read_unreleased_changelog_entries()
+
+    md = compose_release_notes(name, version, entry_list)
+
+    if dry_run:
+        click.echo(md)
+    else:
+        write_release_notes(version, md, overwrite=overwrite)
+
+
+def read_unreleased_changelog_entries():
+    """Import changelog entries to include in the notes."""
+
+    dirpath = determine_changelog_entries_dirpath()
+
+    if not os.path.exists(dirpath):
+        msg = "changelog entries directory {} does not exist.".format(dirpath)
+        raise click.ClickException(msg)
+
+    try:
+        entries = read_changelog_entries(dirpath)
+    except Exception as exc:
+        raise click.ClickException(exc)
+
+    entries = organize_entries_by_category(entries)
+
+    return entries
+
+
+def organize_entries_by_category(entry_list):
+    """Sort entries by category."""
```

Comment:
the docstring of most functions misses the description of the params. Is there a specific reason for that?

Reply:
These functions shouldn't be used by anyone. Only by the script. I also think they are pretty obvious.